File: solutions/3170. Lexicographically Minimum String After Removing Stars/3170.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def clearStars(self, s: str) -> str:
    pos = {chr(i + ord('a')): [] for i in range(26)}
    s = list(s)

    for i , c in enumerate(s):
      if c == '*':
        for p in pos.values():
          if p:
            s[p.pop()] = '*'
            break
      else:
        pos[c].append(i)

    return ''.join(c for c in s if c != '*')

File: solutions/3170. Lexicographically Minimum String After Removing Stars/3170.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
 public:
  string clearStars(string s) {
    vector<int> pos[26];

    for (int i = 0; i < ssize(s); ++i) {
      if (s[i] == '*') {
        for (auto& p : pos) {
          if (!p.empty()) {
            s[p.back()] = '*';
            p.pop_back();
            break;
          }
        }
      } else {
        pos[s[i] - 'a'].push_back(i);
      }
    }
    s.erase(remove(s.begin(), s.end(), '*'), s.end());

    return s;
  }
};